# FCNN Baseline Classifier for Plant Waves
----

#### PREPROCESSING:

In [1]:
import numpy as np
import torch
import pickle
import os
import matplotlib.pyplot as plt
from torch.utils.data import TensorDataset, DataLoader
from tqdm import tqdm

%run ../src/utils/constants.py

Load raw data that has been split into train, test, and validation sets for the 81k and 191k cases.

In [2]:
path_to_pickle_train81k = DATASETS_DIR / "train-81k-stratified-without-neutral.pkl"
path_to_pickle_train191k = DATASETS_DIR / "train-191k-stratified-without-neutral.pkl"
path_to_pickle_test81k = DATASETS_DIR / "test-81k-stratified-without-neutral.pkl"
path_to_pickle_val81k = DATASETS_DIR / "val-81k-stratified-without-neutral.pkl"

In [ ]:
with open(path_to_pickle_train81k, 'rb') as train_81k_file:
    train_81k = pickle.load(train_81k_file)

In [3]:
with open(path_to_pickle_train191k, 'rb') as train_191k_file:
    train_191k = pickle.load(train_191k_file)

In [4]:
with open(path_to_pickle_test81k, 'rb') as test_81k_file:
    test_81k = pickle.load(test_81k_file)

In [5]:
with open(path_to_pickle_val81k, 'rb') as val_81k_file:
    val_81k = pickle.load(val_81k_file)

Just in case you want to use the 191k dataset, run the following line of code. Don't load the `train_81k_file`.

In [6]:
train_81k = train_191k
len(train_81k)

28742

While ignoring the class `Neutral`, separate wave slices and labels, and create a Tensor dataset for train, test, val.

##### Train 81k

In [7]:
wav_slices_train_81k = []
labels_train_81k = []
for segment_train_81k in train_81k:  
    wav_slices_train_81k.append(segment_train_81k[0])
    labels_train_81k.append(segment_train_81k[1])

In [8]:
wav_slices_train_81k = torch.tensor(np.array(wav_slices_train_81k), dtype=torch.float32)
labels_train_81k = torch.tensor(np.array(labels_train_81k), dtype=torch.long)
dataset_train_81k = TensorDataset(wav_slices_train_81k, labels_train_81k)

Normalize the train samples using per-sample normalization.

In [9]:
data_tensor_train_81k, labels_tensor_train_81k = dataset_train_81k.tensors
mean_train_81k = torch.mean(data_tensor_train_81k, dim=1, keepdim=True)
std_dev_train_81k = torch.std(data_tensor_train_81k, dim=1, keepdim=True)
standardized_data_train_81k = (data_tensor_train_81k - mean_train_81k) / (std_dev_train_81k + 0.00000001)

dataset_train_81k = TensorDataset(standardized_data_train_81k, labels_tensor_train_81k)

##### Test 81k

In [10]:
wav_slices_test_81k = []
labels_test_81k = []
for segment_test_81k in test_81k:  
    wav_slices_test_81k.append(segment_test_81k[0])
    labels_test_81k.append(segment_test_81k[1])

In [11]:
wav_slices_test_81k = torch.tensor(np.array(wav_slices_test_81k), dtype=torch.float32)
labels_test_81k = torch.tensor(np.array(labels_test_81k), dtype=torch.long)
dataset_test_81k = TensorDataset(wav_slices_test_81k, labels_test_81k)

Normalize the test samples using per-sample normalization.

In [12]:
data_tensor_test_81k, labels_tensor_test_81k = dataset_test_81k.tensors
mean_test_81k = torch.mean(data_tensor_test_81k, dim=1, keepdim=True)
std_dev_test_81k = torch.std(data_tensor_test_81k, dim=1, keepdim=True)
standardized_data_test_81k = (data_tensor_test_81k - mean_test_81k) / (std_dev_test_81k + 0.00000001)

dataset_test_81k = TensorDataset(standardized_data_test_81k, labels_tensor_test_81k)

##### Validation 81k

In [13]:
wav_slices_val_81k = []
labels_val_81k = []
for segment_val_81k in val_81k:  
    wav_slices_val_81k.append(segment_val_81k[0])
    labels_val_81k.append(segment_val_81k[1])

In [14]:
wav_slices_val_81k = torch.tensor(np.array(wav_slices_val_81k), dtype=torch.float32)
labels_val_81k = torch.tensor(np.array(labels_val_81k), dtype=torch.long)
dataset_val_81k = TensorDataset(wav_slices_val_81k, labels_val_81k)

Normalize the validation samples using per-sample normalization.

In [15]:
data_tensor_val_81k, labels_tensor_val_81k = dataset_val_81k.tensors
mean_val_81k = torch.mean(data_tensor_val_81k, dim=1, keepdim=True)
std_dev_val_81k = torch.std(data_tensor_val_81k, dim=1, keepdim=True)
standardized_data_val_81k = (data_tensor_val_81k - mean_val_81k) / (std_dev_val_81k + 0.00000001)

dataset_val_81k = TensorDataset(standardized_data_val_81k, labels_tensor_val_81k)

#### Create dataloaders

In [16]:
batch_size = 32

In [17]:
train81k_loader = DataLoader(dataset_train_81k, batch_size)
test81k_loader = DataLoader(dataset_test_81k, batch_size)
val81k_loader = DataLoader(dataset_val_81k, batch_size)

#### Hyperparameter Optimization

In [18]:
import wandb
import math
import optuna
from optuna.trial import TrialState
from sklearn.metrics import f1_score, accuracy_score, balanced_accuracy_score, precision_score, recall_score

C:\Users\nekta\OneDrive\Desktop\botanic-emotion-recognition\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Define the FCNN model.

In [19]:
def define_model(trial):
    layers = []

    INPUT_DIM = 10000
    N_CLASSES = 6
    dropout_rate = trial.suggest_categorical("dropout_rate", [0.0, 0.1, 0.2])

    hidden_dim_1 = trial.suggest_categorical("hidden_dim_1", [2 ** i for i in range(4, 8)])

    layers.append(torch.nn.Linear(INPUT_DIM, hidden_dim_1))
    layers.append(torch.nn.ReLU())
    layers.append(torch.nn.Dropout(dropout_rate))

    hidden_dim_2 = trial.suggest_categorical("hidden_dim_2", [2 ** i for i in range(4, 7)])
    
    layers.append(torch.nn.Linear(hidden_dim_1, hidden_dim_2))
    layers.append(torch.nn.ReLU())
    layers.append(torch.nn.Dropout(dropout_rate))
    
    layers.append(torch.nn.Linear(hidden_dim_2, N_CLASSES))

    config_dict = {
        "hidden_dim_1": hidden_dim_1,
        "hidden_dim_2": hidden_dim_2,
        "dropout_rate": dropout_rate
    }

    return torch.nn.Sequential(*layers), config_dict

Define the objective function for optuna for the hyperparameter optimization.

In [20]:
def objective(trial): 
    model, config_dict = define_model(trial)

    lr = trial.suggest_categorical('lr', [0.0001, 0.001, 0.01])

    criterion = torch.nn.CrossEntropyLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=lr)

    print(config_dict)
    name_experiment = (f"{trial.number}_{NAME_CORE}_lr-{lr}_hidden1-{config_dict['hidden_dim_1']}_hidden2-"
                       f"{config_dict['hidden_dim_2']}_dr-{config_dict['dropout_rate']}")

    config_dict["lr"] = lr
    config_dict["epochs"] = NUM_EPOCHS

    wandb.init(
        project="model_" + NAME_CORE + "-hpo",
        dir=LOGS_DIR,
        name=name_experiment,
        config=config_dict
    )


    for epoch in range(NUM_EPOCHS):
        model.train()
        for batch_data, batch_labels in tqdm(train81k_loader):          
            # zero gradients
            optimizer.zero_grad()

            # forward + backward pass
            outputs = model(batch_data)
            loss = criterion(outputs, batch_labels)
            loss.backward()
            optimizer.step()

        all_preds = []
        all_labels = []

        model.eval()
        with torch.no_grad():
            for batch_data, batch_labels in tqdm(val81k_loader): 
                output = model(batch_data)

                predicted = output.argmax(dim=1)
                all_preds.extend(predicted.numpy())
                all_labels.extend(batch_labels.numpy())

        # TODO: explicitely state in written thesis that zero_division=0.0
        balanced_accuracy = balanced_accuracy_score(all_labels, all_preds)
        accuracy = accuracy_score(all_labels, all_preds)
        f1_class = f1_score(all_labels, all_preds, average=None, zero_division=0.0)
        f1_weighted = f1_score(all_labels, all_preds, average="weighted", zero_division=0.0)
        recall = recall_score(all_labels, all_preds, average="weighted", zero_division=0.0)
        precision = precision_score(all_labels, all_preds, average="weighted", zero_division=0.0)

        trial.report(balanced_accuracy, epoch)

        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()

        metrics = {
            "balanced_accuracy": balanced_accuracy,
            "accuracy": accuracy,
            "f1_weighted": f1_weighted,
            "recall": recall,
            "precision": precision
        }

        wandb_input = metrics
        wandb.log(wandb_input)

    wandb.finish()

    return balanced_accuracy

Run hyperparameter optimization using optuna. Specify a study name and define the search space for the grid search. 

In [21]:
search_space = {
    'lr': [0.0001, 0.001, 0.01],
    'hidden_dim_1': [2 ** i for i in range(4, 8)],
    'hidden_dim_2': [2 ** i for i in range(4, 7)],
    "dropout_rate": [0, 0.1, 0.2]
}

In [22]:
num_combi = 1
for i in search_space:
    num_combi *= len(search_space[i])
num_combi

108

In [23]:
NAME_CORE = f"fcnn_baseline_6_large_81k_search_{num_combi}"
NUM_EPOCHS = 20

#### Second Run:

In [24]:
sampler = optuna.samplers.GridSampler(search_space)
study = optuna.create_study(sampler=sampler, study_name=NAME_CORE, storage="sqlite:///hpo_" + NAME_CORE + ".db",
                            direction="maximize", load_if_exists=True)
study.optimize(objective, n_trials=num_combi)

[I 2024-02-14 20:49:20,989] A new study created in RDB with name: fcnn_baseline_6_large_81k_search_108


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0}


wandb: Currently logged in as: totikos. Use `wandb login --relogin` to force relogin


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 702.89it/s]


[I 2024-02-14 20:51:08,262] Trial 0 finished with value: 0.16666666666666666 and parameters: {'dropout_rate': 0.0, 'hidden_dim_1': 16, 'hidden_dim_2': 64, 'lr': 0.01}. Best is trial 0 with value: 0.16666666666666666.


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 471.24it/s]


[I 2024-02-14 21:01:09,515] Trial 1 finished with value: 0.16666666666666666 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 128, 'hidden_dim_2': 32, 'lr': 0.01}. Best is trial 0 with value: 0.16666666666666666.


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 649.72it/s]


[I 2024-02-14 21:03:12,925] Trial 2 finished with value: 0.16666666666666666 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 16, 'hidden_dim_2': 64, 'lr': 0.01}. Best is trial 0 with value: 0.16666666666666666.


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 654.68it/s]


[I 2024-02-14 21:04:25,980] Trial 3 finished with value: 0.16567810404743757 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 16, 'hidden_dim_2': 64, 'lr': 0.001}. Best is trial 0 with value: 0.16666666666666666.


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 721.22it/s]


[I 2024-02-14 21:07:25,940] Trial 4 finished with value: 0.16666666666666666 and parameters: {'dropout_rate': 0.0, 'hidden_dim_1': 32, 'hidden_dim_2': 64, 'lr': 0.01}. Best is trial 0 with value: 0.16666666666666666.


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 461.12it/s]


[I 2024-02-14 21:11:31,394] Trial 5 finished with value: 0.16672857608341482 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 128, 'hidden_dim_2': 32, 'lr': 0.001}. Best is trial 5 with value: 0.16672857608341482.


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 576.69it/s]


[I 2024-02-14 21:14:03,193] Trial 6 finished with value: 0.17126425545780385 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 64, 'hidden_dim_2': 16, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 574.85it/s]


[I 2024-02-14 21:15:56,490] Trial 7 finished with value: 0.16604956054792633 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 32, 'hidden_dim_2': 32, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 900.85it/s]
[I 2024-02-14 21:16:00,758] Trial 8 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 283.00it/s]
[I 2024-02-14 21:16:27,892] Trial 9 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 478.28it/s]
[I 2024-02-14 21:16:54,330] Trial 10 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 570.10it/s]


[I 2024-02-14 21:19:00,465] Trial 11 finished with value: 0.16851411658147483 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 32, 'hidden_dim_2': 16, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 573.85it/s]
[I 2024-02-14 21:19:08,898] Trial 12 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 476.38it/s]
[I 2024-02-14 21:19:29,012] Trial 13 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 809.36it/s]
[I 2024-02-14 21:19:38,992] Trial 14 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 579.34it/s]
[I 2024-02-14 21:19:52,392] Trial 15 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 507.02it/s]
[I 2024-02-14 21:20:11,958] Trial 16 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 655.49it/s]
[I 2024-02-14 21:20:24,223] Trial 17 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 390.41it/s]
[I 2024-02-14 21:20:45,167] Trial 18 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 596.87it/s]


[I 2024-02-14 21:23:20,807] Trial 19 finished with value: 0.17029866664370005 and parameters: {'dropout_rate': 0.0, 'hidden_dim_1': 64, 'hidden_dim_2': 32, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 855.48it/s]
[I 2024-02-14 21:23:25,079] Trial 20 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 683.12it/s]
[I 2024-02-14 21:23:38,097] Trial 21 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 525.36it/s]


[I 2024-02-14 21:26:23,899] Trial 22 finished with value: 0.16652003910068425 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 64, 'hidden_dim_2': 32, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 828.82it/s]
[I 2024-02-14 21:26:28,189] Trial 23 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 431.45it/s]


[I 2024-02-14 21:30:41,182] Trial 24 finished with value: 0.1629308934161854 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 128, 'hidden_dim_2': 32, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 703.07it/s]
[I 2024-02-14 21:30:47,449] Trial 25 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 528.96it/s]


[I 2024-02-14 21:33:28,732] Trial 26 finished with value: 0.1619485174323884 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 64, 'hidden_dim_2': 16, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 409.86it/s]
[I 2024-02-14 21:33:41,481] Trial 27 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 745.80it/s]
[I 2024-02-14 21:33:53,205] Trial 28 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 555.29it/s]


[I 2024-02-14 21:36:39,480] Trial 29 finished with value: 0.1653296213520741 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 64, 'hidden_dim_2': 32, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 720.58it/s]
[I 2024-02-14 21:36:45,495] Trial 30 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 777.27it/s]
[I 2024-02-14 21:36:57,205] Trial 31 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 714.69it/s]


[I 2024-02-14 21:38:15,225] Trial 32 finished with value: 0.1640550667970023 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 16, 'hidden_dim_2': 32, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 681.82it/s]
[I 2024-02-14 21:38:21,294] Trial 33 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 912.57it/s]
[I 2024-02-14 21:38:31,205] Trial 34 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 594.73it/s]
[I 2024-02-14 21:38:46,383] Trial 35 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 437.23it/s]
[I 2024-02-14 21:39:05,650] Trial 36 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 459.20it/s]
[I 2024-02-14 21:39:23,880] Trial 37 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 425.30it/s]
[I 2024-02-14 21:39:41,146] Trial 38 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 726.31it/s]
[I 2024-02-14 21:39:53,178] Trial 39 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 779.64it/s]
[I 2024-02-14 21:40:04,812] Trial 40 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 562.88it/s]


[I 2024-02-14 21:42:40,255] Trial 41 finished with value: 0.1666096448354513 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 64, 'hidden_dim_2': 16, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 714.96it/s]
[I 2024-02-14 21:42:46,189] Trial 42 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 869.95it/s]
[I 2024-02-14 21:42:56,097] Trial 43 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 728.52it/s]
[I 2024-02-14 21:43:07,401] Trial 44 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 619.99it/s]
[I 2024-02-14 21:43:18,738] Trial 45 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 562.16it/s]
[I 2024-02-14 21:43:32,544] Trial 46 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 451.23it/s]
[I 2024-02-14 21:43:52,736] Trial 47 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 821.77it/s]
[I 2024-02-14 21:44:03,048] Trial 48 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 420.87it/s]
[I 2024-02-14 21:44:20,948] Trial 49 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 856.00it/s]
[I 2024-02-14 21:44:32,674] Trial 50 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 863.41it/s]
[I 2024-02-14 21:44:44,839] Trial 51 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 692.85it/s]
[I 2024-02-14 21:44:58,424] Trial 52 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 951.94it/s]
[I 2024-02-14 21:45:12,637] Trial 53 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 444.89it/s]
[I 2024-02-14 21:45:30,584] Trial 54 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 700.49it/s]


[I 2024-02-14 21:46:49,026] Trial 55 finished with value: 0.16833059886918067 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 16, 'hidden_dim_2': 16, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 329.10it/s]
[I 2024-02-14 21:47:02,816] Trial 56 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 463.14it/s]
[I 2024-02-14 21:47:18,593] Trial 57 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 473.53it/s]
[I 2024-02-14 21:47:29,497] Trial 58 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 518.73it/s]


[I 2024-02-14 21:50:16,849] Trial 59 finished with value: 0.17065507282574213 and parameters: {'dropout_rate': 0.0, 'hidden_dim_1': 64, 'hidden_dim_2': 64, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 436.11it/s]
[I 2024-02-14 21:50:28,913] Trial 60 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 651.45it/s]
[I 2024-02-14 21:50:43,263] Trial 61 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 626.35it/s]


[I 2024-02-14 21:52:42,393] Trial 62 finished with value: 0.16819429910875128 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 32, 'hidden_dim_2': 16, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 64, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 554.03it/s]
[I 2024-02-14 21:52:51,094] Trial 63 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 449.91it/s]
[I 2024-02-14 21:53:08,904] Trial 64 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 546.69it/s]
[I 2024-02-14 21:53:22,884] Trial 65 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 851.92it/s]
[I 2024-02-14 21:53:32,642] Trial 66 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 607.36it/s]


[I 2024-02-14 21:54:49,258] Trial 67 finished with value: 0.16693269199024494 and parameters: {'dropout_rate': 0.0, 'hidden_dim_1': 16, 'hidden_dim_2': 64, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 528.67it/s]


[I 2024-02-14 21:57:27,252] Trial 68 finished with value: 0.16917309373398617 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 64, 'hidden_dim_2': 64, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 619.63it/s]
[I 2024-02-14 21:57:35,587] Trial 69 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 660.60it/s]


[I 2024-02-14 21:58:58,127] Trial 70 finished with value: 0.16666666666666666 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 16, 'hidden_dim_2': 16, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 832.89it/s]
[I 2024-02-14 21:59:02,515] Trial 71 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 433.71it/s]


[I 2024-02-14 22:03:17,720] Trial 72 finished with value: 0.16666666666666666 and parameters: {'dropout_rate': 0.2, 'hidden_dim_1': 128, 'hidden_dim_2': 16, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 520.17it/s]
[I 2024-02-14 22:03:25,868] Trial 73 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 477.94it/s]
[I 2024-02-14 22:03:40,084] Trial 74 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 542.92it/s]
[I 2024-02-14 22:03:54,111] Trial 75 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 417.84it/s]
[I 2024-02-14 22:04:17,483] Trial 76 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 760.12it/s]
[I 2024-02-14 22:04:30,436] Trial 77 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 772.68it/s]
[I 2024-02-14 22:04:40,602] Trial 78 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 762.51it/s]
[I 2024-02-14 22:04:50,148] Trial 79 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 423.82it/s]
[I 2024-02-14 22:05:08,896] Trial 80 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 686.27it/s]
[I 2024-02-14 22:05:20,329] Trial 81 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 421.84it/s]
[I 2024-02-14 22:05:38,135] Trial 82 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 710.51it/s]
[I 2024-02-14 22:05:49,333] Trial 83 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 408.12it/s]
[I 2024-02-14 22:06:08,345] Trial 84 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 711.47it/s]
[I 2024-02-14 22:06:20,321] Trial 85 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 534.42it/s]
[I 2024-02-14 22:06:33,462] Trial 86 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 697.19it/s]
[I 2024-02-14 22:06:46,758] Trial 87 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 717.76it/s]
[I 2024-02-14 22:06:58,411] Trial 88 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 513.92it/s]
[I 2024-02-14 22:07:11,973] Trial 89 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 644.37it/s]


[I 2024-02-14 22:08:37,097] Trial 90 finished with value: 0.16367383512544803 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 16, 'hidden_dim_2': 16, 'lr': 0.001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 128, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 420.31it/s]
[I 2024-02-14 22:08:50,832] Trial 91 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 715.28it/s]
[I 2024-02-14 22:09:04,629] Trial 92 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 32, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 504.72it/s]
[I 2024-02-14 22:09:18,080] Trial 93 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 572.36it/s]
[I 2024-02-14 22:09:33,600] Trial 94 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 518.50it/s]
[I 2024-02-14 22:09:52,103] Trial 95 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 455.12it/s]
[I 2024-02-14 22:10:11,059] Trial 96 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 467.62it/s]


[I 2024-02-14 22:14:34,460] Trial 97 finished with value: 0.16190000043062427 and parameters: {'dropout_rate': 0.0, 'hidden_dim_1': 128, 'hidden_dim_2': 64, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 744.78it/s]
[I 2024-02-14 22:14:38,976] Trial 98 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 32, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 822.38it/s]
[I 2024-02-14 22:14:49,271] Trial 99 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 821.20it/s]
[I 2024-02-14 22:15:00,616] Trial 100 pruned. 


{'hidden_dim_1': 16, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 910.11it/s]
[I 2024-02-14 22:15:13,589] Trial 101 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 16, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 634.50it/s]
[I 2024-02-14 22:15:27,075] Trial 102 pruned. 


{'hidden_dim_1': 128, 'hidden_dim_2': 64, 'dropout_rate': 0.2}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 426.95it/s]
[I 2024-02-14 22:15:46,899] Trial 103 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 317.78it/s]
[I 2024-02-14 22:15:58,551] Trial 104 pruned. 


{'hidden_dim_1': 32, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 661.41it/s]


[I 2024-02-14 22:18:01,509] Trial 105 finished with value: 0.16763639665089147 and parameters: {'dropout_rate': 0.1, 'hidden_dim_1': 32, 'hidden_dim_2': 64, 'lr': 0.0001}. Best is trial 6 with value: 0.17126425545780385.


{'hidden_dim_1': 32, 'hidden_dim_2': 16, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 745.28it/s]
[I 2024-02-14 22:18:07,683] Trial 106 pruned. 


{'hidden_dim_1': 64, 'hidden_dim_2': 64, 'dropout_rate': 0.1}


100%|█████████████████████████████████████████| 82/82 [00:00<00:00, 490.61it/s]
[I 2024-02-14 22:18:21,259] Trial 107 pruned. 


In [25]:
pruned_trials = study.get_trials(deepcopy=False, states=[TrialState.PRUNED])
complete_trials = study.get_trials(deepcopy=False, states=[TrialState.COMPLETE])

In [26]:
print("Study statistics: ")
print("  Number of finished trials: ", len(study.trials))
print("  Number of complete trials: ", len(complete_trials))
print("  Number of pruned trials: ", len(pruned_trials))

print("Best trial:")
trial_ = study.best_trial

print("  Value: ", trial_.value)
print("  Params: ")
for key, value in trial_.params.items():
    print("    {}: {}".format(key, value))

Study statistics: 
  Number of finished trials:  108
  Number of complete trials:  26
  Number of pruned trials:  82
Best trial:
  Value:  0.17126425545780385
  Params: 
    dropout_rate: 0.2
    hidden_dim_1: 64
    hidden_dim_2: 16
    lr: 0.001


#### 191k RUN 
Study statistics: 
  Number of finished trials:  108
  Number of complete trials:  26
  Number of pruned trials:  82
Best trial:
  Value:  0.17126425545780385
  Params: 
    dropout_rate: 0.2
    hidden_dim_1: 64
    hidden_dim_2: 16
    lr: 0.001

#### First Run:

In [ ]:
sampler = optuna.samplers.GridSampler(search_space)
study = optuna.create_study(sampler=sampler, study_name=NAME_CORE, storage="sqlite:///hpo_" + NAME_CORE + ".db",
                            direction="maximize", load_if_exists=True)
study.optimize(objective, n_trials=num_combi)

Closer look into the trials. 

In [ ]:
pruned_trials = study.get_trials(deepcopy=False, states=[TrialState.PRUNED])
complete_trials = study.get_trials(deepcopy=False, states=[TrialState.COMPLETE])

In [ ]:
print("Study statistics: ")
print("  Number of finished trials: ", len(study.trials))
print("  Number of complete trials: ", len(complete_trials))
print("  Number of pruned trials: ", len(pruned_trials))

print("Best trial:")
trial_ = study.best_trial

print("  Value: ", trial_.value)
print("  Params: ")
for key, value in trial_.params.items():
    print("    {}: {}".format(key, value))

#### Training the Baseline FCNN.
10 times training with the best HP from HPO, and then average. 

#### Testing the Baseline Classifier

In [ ]:
import os
import numpy as np
from tqdm import tqdm
from PIL import Image
import matplotlib.pyplot as plt

%run ../src/utils/constants.py

In [ ]:
test_81k_folder = "test-81k"